### PR TITLE
Allow integration with Hardare Security Modules

### PR DESCRIPTION
**Pull request by JaimieW**
Added methods which can be overridden to allow integration with Hardware Security Modules
I added issue #1291 for this

### CODE REVIEW

**Comment by boaks** [comment]
Thanks for your contribution!
Please read [CONTRIBUTING](https://github.com/eclipse/californium/blob/master/CONTRIBUTING.md)

Especially, the ECA is required. Please use the same e-mail for registering there as you use for your commits. Finally the commit needs a "sign", which is just a text as:

```
Signed-off-by: Achim Kraus <achim.kraus@bosch.io>
```

**Comment by JaimieW** [comment]
Hi,
I have read the Contributing page.
I have signed the ECA.
Do I just add the text to the commit message?

Thanks,
Jaimie
________________________________

**Comment by boaks** [comment]
yes ... then force push ... and it should go green ...

**Comment by boaks** [comment]
Please take care, that you use the same e-mail for the commits, sign and the ECA.

**Comment by boaks** [comment]
> I have signed the ECA.

At least the "check tool" reports, that you didn't sign a ECA with the e-mail in the commits.

**Comment by JaimieW** [comment]
Sorry, I'm new to open source contributing.

Is there a particular format for
Signed-off-by: Jaimie Whiteside <jaimiewhiteside@hotmail.com>

My github and eclipse emails are the same.
I'm not sure what I'm doing wrong.

**Comment by boaks** [comment]
Your commits shows an other e-mail as "author" and "commiter" then the signing and the ECA.

"jaimiewhiteside@hotmail.com" has a valid ECA, that's great.

Now create commits with that as author.

`git commit --amend --no-edit --reset-author`

would do that, if the e-mail config would be the hotmail and not the business mail.

**Comment by JaimieW** [comment]
Got it now updated now, thank you.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java [hunk 1]
```diff
@@ -885,12 +885,23 @@ protected void calculateKeys(SecretKey masterSecret) {
...
 	}
+	
+	/**
+	 * Derives the master secret from the premaster secret and the random seed.
+	 * 
+	 * @param premasterSecret the Pre Master Secret Key
+	 * @param randomSeed the random seed bytes
+	 * @return the derived master secret bytes
+	 */
+	protected byte[] deriveMasterSecret(SecretKey premasterSecret, byte[] randomSeed) {
```
Comment (boaks) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/Handshaker.java:901]: Did you have a look at the [SecretUtil](https://github.com/eclipse/californium/blob/master/scandium-core/src/main/java/org/eclipse/californium/scandium/util/SecretUtil.java)? Maybe you can adapt the return type to a `SecretKey`?

**Comment by boaks** [comment]
Generally, scandium uses some "callback" interfaces, e.g. [x509 - CertificateVerifier](https://github.com/eclipse/californium/blob/master/scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/x509/CertificateVerifier.java) to make such functions configurable. To permit to use a own `Handshaker` subclass will lock more functionality as "public API" and would make it harder to change scandium for maintenance.
 
If you would have a look on such an approach, I would appreciate it.
Otherwise, we can merge it for now. But, though the async PSK API is on my list, I may change that again before release a 2.3.

**Comment by JaimieW** [comment]
Added a commit to use an interface to derive the master secret in the Handshaker.
If it is not configured in the DtlsConnectorConfig, then the default behaviour is to use the existing PseudoRandomFunction static method to derive the master secret.

**Comment by boaks** [comment]
The interface looks very good! I prefer that way.

Currently I'm wondering, if your HSM stores only one psk?
Or do you encode the identity in the returned key of `PskStore.getKey()`?
What happens, if a `premasterSecret` of a ECDSA handshake is provided for the call?

**Comment by JaimieW** [comment]
My implementation only supports PSK, this is because I have to be compliant with an internal specification.
Therefore I haven't tried an ECDSA handshake.

I use the HSM to store the PSK for each client that will connect to the server.
The client's unique identifer is transmitted in the psk id. 
My custom PskStore uses the identifierto lookup and return the corresponding PSK key reference from the HSM.
My custom deriver class is a wrapper, which gets the HSM to derive the master secret using the PSK reference.

**Comment by boaks** [comment]
OK, now I understand better how it works.
Assuming the interface would be a 

```
interface HsmPskStore {
SecretKey getMasterKey(ServerNames serverName, PskPublicInformation identity, Mac hmac, byte[] randomSeed);
SecretKey getMasterKey(PskPublicInformation identity, Mac hmac, byte[] randomSeed);
}
```

and implementing that as e.g.

```
class XyzHsmPskStore implements PskStore, HsmPskStore {
...
}
```

Do you think, that will also possible? It would require some more internal changes in scandium, so I would merge your PR "as it is", and adapt the implementation in the next two or three weeks.

**Comment by JaimieW** [comment]
I'm not sure I fully understand your intention with the HsmPskStore interface.
Would the call to getMasterKey be made from receivedClientKeyExchange?
Is it a replacement for the sequence:
```
SecretKey premaster = pskUtil.generatePremasterSecretFromPSK(null);
...
generateKeys(premasterSecret)
```


There may be a problem in that the HSM client may not conform to the JCE api, so there may not be a valid Mac object.

**Comment by boaks** [comment]
> Would the call to getMasterKey be made from receivedClientKeyExchange?
> Is it a replacement for the sequence:
> 

Yes, after some additional adaption.

The `MAC` is passed in empty. You may just use the "getAlgorithm" to check, which MAC to use on the HSM (or the cipher suite doesn't match your HSM). There is no need to call update or doFinal on it, just use the algorithm to calculate the master secret.

**Comment by boaks** [comment]
Just to write it again:

For me, this PR is "LGTM".
I just ask the questions in order to ensure, that such a solution works also for you. It will be my task to change scandium accordingly.

**Comment by JaimieW** [comment]
Yes, fine by me.On 24 Apr 2020 14:51, Achim Kraus <notifications@github.com> wrote:
Just to write it again:
For me, this PR is "LGTM".
I just ask the questions in order to ensure, that such a solution works also for you. It will be my task to change scandium accordingly.

**Comment by boaks** [comment]
LGTM

**Comment by sbernard31** [comment]
I had a quick look : no objection :)

**Comment by boaks** [comment]
@sbernard31 

Just to mention it:
I started to work on the async PSK store at the weekend.
My preference is currently, to offer a interface, which returns the master secret asynchronous, not only the secret key. If someone only wants to fetch the secret key asynchronous, then it will be very easy to use that to provide the master secret and use then the asynchronous handler. My current design is "stateless", I use the CID as ID to match the master secret for the handshakes.
Generally, the idea already works, but need more work. Hope I can finish my exploration this week or next weekend to make a first PR to get more feedback.

**Comment by sbernard31** [comment]
@boaks  :+1:

**Pull request merged**